File: leetcode-34-search-for-a-range/solution.py
#!/usr/bin/python
#-*- coding:utf-8 -*-
#================================================================
#  Copyright (C) 2020 Li Mingyang. All rights reserved.
#  
#  FileName : solution.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        ret = [-1, -1]

        left, right = 0, len(nums)-1
        while left<=right:
            mid = (left+right)/2
            item = nums[mid]
            if item<target:
                left += 1
            elif item>=target:
                right -= 1
                if item==target:
                    ret[0] = mid
        
        left, right = 0, len(nums)-1
        while left<=right:
            mid = (left+right)/2
            item = nums[mid]
            if item>target:
                right -= 1
            elif item<=target:
                left += 1
                if item==target:
                    ret[1] = mid
        
        return ret

File: leetcode-34-search-for-a-range/solution.cpp
/*================================================================
*  Copyright (C) 2020 Limingyang. All rights reserved.
*  
*  FileName : solution.cpp
*    Author : Limingyang
*      Date : 2020-11-29
*      Brif : 
*
================================================================*/
#include "solution.h"
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
    
        int size = nums.size();
        vector<int> ret(2,-1);

        // 寻找左端点
        for(int left=0, right=size-1; left<=right; ){
            int mid = (left+right)/2;
            int n = nums[mid];

            if (n<target){  // n target 
                left = mid+1;
            }else if(n>=target){  // target n 
                right = mid-1;
                if (n==target){
                    ret[0] = mid;
                }
            }
        }

        // 寻找右端点
        for(int left=0, right=size-1; left<=right; ){
            int mid = (left+right)/2;
            int n = nums[mid];

            if (n>target){  // target n 
                right = mid-1;
            }else if(n<=target){  // n target 
                left = mid+1;
                if (n==target){
                    ret[1] = mid;
                }
            }
        }

        return ret;

    }
};
